### DISCUSSION TITLE
Extending user plugin - get data from other table

### DISCUSSION THREAD

**Discussion by AIC-BV** [wintercms/winter#373] [Q&A] [answered]
Hi

I want to extend Winter.User, by adding links to the user order(s).
The orders are stored in a seperate table, but have a column 'user_id':

**QUERY (not yet used)**
```php
$orders = Db::table('aic_account_orders')
                        ->where('user_id', $user->id)
                        ->latest()
                        ->get();
```

**extenduser PLUGIN.PHP**
```php
public function boot()
{
        UserModel::extend(function ($model) {
             $model->hasMany['orders'] = ['Aic\Account\Models\Orders'];
        });

        UsersController::extendFormFields(function ($form, $model, $context) {
                $form->addTabFields([
                        'orders' => [
                                'tab' => 'Orders',
                                'type' => 'partial',
                                'path' => 'my_path'
                        ]
                ]);
        }
}
```

I want to create a list of links to his/her order(s).
I would do this by getting all the orders.
Then I can create my own url in a partial with a foreach loop on orders: '/orders/preview/' . $id

**The wanted result**
```php
<ul>
    <li><a href='my_generated_link_to_the_order' target='_blank'>my_order_name</a></li>
    <li><a href='my_generated_link_to_the_order' target='_blank'>my_order_name</a></li>
</ul>
```

How can I populate the 'orders' field with the correct data from the DB?
Is this possible? Can you point me in the right direction?

Thanks!

**EDIT**
$value in my partial returns an array of IDs but I have no additional information (but can build hardcoded urls).
```php
<?php $url = 'http://127.0.0.1:8000/cms/aic/account/orders/preview/'; ?>
<ul>
    <?php foreach ($value as $id) {?>
        <li><a href='<?= e($url . $id) ?>' target='_blank'><?= e($url . $id) ?></a></li>
    <?php }?>
</ul>
```

**Comment by LukeTowers** [comment, marked as answer]
The best thing that you can do is use the RelationController functionality: https://wintercms.com/docs/backend/relations. You will need to extend the user plugin to add it dynamically to the Users controller and configure it to display your orders relationship correctly, but it will provide an embedded list with all of the user's orders and clickable rows that can either popup the order in a modal or be configured to take you to the preview page for the order. 

If you do just want to go down the simple partial route then you can just reference the `$model` variable inside of your partial (see https://wintercms.com/docs/backend/forms#field-partial) and call the orders relationship directly on that:

```php
<?php if ($model->orders->count()) { ?>
<ul>
    <?php foreach ($model->orders as $order) {?>
        <li><a href='<?= e(\Backend::url("aic/account/orders/preview/{$order->id}")); ?>' target='_blank'><?= e($order->name); ?></a></li>
    <?php } ?>
</ul>
<?php } ?>

**Reply by AIC-BV** [reply]
Thank you!!

The simple partial way works perfectly! 😍
I think I can do everything I want, and more, with the model data.

However, I am intrigued to try out the RelationController 😁

**Reply by AIC-BV** [reply]
The RelationController is awesome!
It's much better than the simple partial solution! 😄

-----------------------

Noticed 2 bugs(?):
1. recordsPerPage is set on 2 but it still displays all my records (3)
**(SOLVED)** I don't know why the docs put recordsPerPage at manage, but when I add it to 'view' it works
```
manage:
        form:  ~/plugins/aic/account/models/orders/fields.yaml
        recordsPerPage: 2
```

2. When clicking the record, secondaryTabs are displayed on every tab at the bottom labelled as 'misc'.

-----------------------

And I have one issue; I can't find how to remove a column
```
UserModel::extend(function ($model) {

            // remove columns
            $model->addDynamicMethod('relationExtendViewWidget', function() use ($model) {
                // // Make sure the model and field matches those you want to manipulate
                // if (!$model instanceof MyModel || $field != 'myField') return;
                
                $model->bindEvent('list.extendColumns', function() use($model) {
                    $model->removeColumn('user_id');
                });
            });

});
```

**Reply by LukeTowers** [reply]
2. Secondary tabs are displayed that way by default, if you want to change how the form is rendered then you can override the partials using the instructions here: https://wintercms.com/docs/backend/relations#override-relation-partials

Your issue with removing a column is because the `relationExtendViewWidget` isn't a dynamic method that gets added to a model, it's a real method that you add to your controller. Since dynamic methods can't overwrite existing methods, you may not be able to dynamically add it to the Users controller, but the good news is you don't have to.

Instead of referring to the user plugin's columns.yaml in your relation configuration just create your own columns.yaml and reference that instead.

**Reply by AIC-BV** [reply]
> Instead of referring to the user plugin's columns.yaml in your relation configuration just create your own columns.yaml and reference that instead.

Yes I already did that 😁 But I'm glad you confirmed that this is the way to go. 😊
Thanks for your time and help!
